### ISSUE TITLE
Odd scoping in refinement types

### ISSUE THREAD

**Issue by sdleffler** [ucsd-progsys/liquidhaskell#1074]
I'm working with the function `the`, as defined here:
```haskell
the :: Eq a => [a] -> a
the (x:xs)
  | null (filter (/= x) xs) = x
  | otherwise               = errorWithoutStackTrace "GHC.Exts.the: non-identical elements"
the []                      = errorWithoutStackTrace "GHC.Exts.the: empty list"
```
I have a type alias I defined for working with nonempty lists, `NEList a`, defined as
```haskell
type NEList a = {xs:[a] | len xs > 0}
```
My problem goes something like this - I wanted to give `the` the following refinement type:
```haskell
{-@ the :: Eq a => xs:NEList {x:a | x == headP xs} -> {y:a | headP xs == y} @-}
```
However, this fails. The refinement type passed into `NEList` is unable to refer to `xs` outside:
```haskell
 sean  …  ghc  libraries  base  liquidbase GHC/Exts.hs 
LiquidHaskell Copyright 2013-17 Regents of the University of California. All Rights Reserved.


**** DONE:  A-Normalization ****************************************************
 

**** DONE:  annotate ***********************************************************
 

**** RESULT: ERROR *************************************************************


 /home/sean/Projects/Haskell/ghc/libraries/base/GHC/Exts.hs:118:12: Error: Illegal type specification for `GHC.Exts.the`
 
 118 | {-@ the :: Eq a => xs:NEList {x:a | x == headP xs} -> {y:a | headP xs == y} @-}
                  ^
     GHC.Exts.the :: (Eq a) =>
                     xs:{v : [{VV : a | VV == headP xs}] | len v > 0} -> {VV : a | headP xs == VV}
     Sort Error in Refinement: {VV : a_a7TB2 | VV == headP xs}
     Unbound Symbol xs
 Perhaps you meant: VV, fst
 sean  …  ghc  libraries  base  
```
However, when I rewrite the refinement type for `the` as follows:
```haskell
{-@ the :: Eq a => {xs:[{x:a | x == headP xs}] | len xs > 0} -> {y:a | headP xs == y} @-}
```
The type specification becomes legal.

The problem does not seem to lie with the type alias - the following refinement type definition produces the exact same error:
```haskell
{-@ the :: Eq a => xs:{v:[{x:a | x == headP xs}] | len v > 0} -> {y:a | headP xs == y} @-}
```
Why does LH consider this specification different from the prior?

**Comment by ranjitjhala** [comment]
Hi, yes the binder "inside" the refinement is allowed but the one "outside"
is really a function parameter binder and is in scope at the output type,
which is what you're seeing.

That said, you may be right, perhaps it makes sense to treat

x:{v:t | p }

Just the same as

x:{x:t | p[x/v] }

This is "probably" easy to do... :)

**Comment by sdleffler** [comment]
Huh! I was under the impression that LH already treated them identically, since for both `xs:{...}` and `{xs:...}`, `xs` remains in scope after the binder. Otherwise the specification
```haskell
{-@ the :: Eq a => {xs:[{x:a | x == headP xs}] | len xs > 0} -> {y:a | headP xs == y} @-}
```
wouldn't pass muster. Is this unintentional?

**Comment by ranjitjhala** [comment]
No this is intentional, but the way it works is that

 `{xs:...} -> thing`

is "hoisted outside" as a function binder wherever appropriate
(i.e. in a function signature, without an explicit binder), and
hence treated the same as

 `xs:{xs:...} -> thing`

instead of what you're suggesting which is the other way around
i.e. pushing the "outside" function binder "inside" the refinement.

**Comment by Gabriel439** [comment]
My intuition has always been that in:

```
x : { y : a | pred } -> b
```

`x` should be in scope for `b` and not `a` nor `pred` (consistent with other dependently typed languages where bound variables are not in scope for their own type)

`y` should be in scope for `pred` and not `a` nor `b` (i.e. the scope of `y` is the braces)

**Comment by ranjitjhala** [comment]
Hi @sdleffler thanks again for bringing this up. 
I hit this myself right now in another fashion, namely, that as the function binder is not in 
scope, we cannot use type aliases with function binders.

In short, the [following fails](http://goto.ucsd.edu:8090/index.html#?demo=permalink%2F1504287805_3396.hs) with a similar error `y is not in scope`...

```haskell
{-@ type Geq X = {v:Int | X <= v} @-} 

{-@ foo :: x:Int -> y:Geq y -> () @-}
foo :: Int -> Int -> ()
foo x y = ()
```

The above bothers me especially 
as a I like to use aliases! 

Let me see if there is a fix...

**Comment by ranjitjhala** [comment]
Even more confusing:

```haskell
module Blank where

{-@ type Geq X = {v:Int | X <= v} @-} 

{-@ thisOk :: x:Int -> y:{Int | y > 10}  -> () @-}
thisOk :: Int -> Int -> ()
thisOk x y = ()

{-@ thisBad1 :: x:Int -> y:Geq y  -> () @-}
thisBad1 :: Int -> Int -> ()
thisBad1 x y = ()

{-@ thisBad2 :: x:Int -> y:{y > 10}  -> () @-}
thisBad2 :: Int -> Int -> ()
thisBad2 x y = ()
```

**Comment by ranjitjhala** [comment]
Here's some quirks in the parsing:

```haskell
λ> doParse' bareTypeP "" "x:{0 < x} -> Int"
x:{VV : _ | 0 < x} -> Int
λ> doParse' bareTypeP "" "x:{Int | 0 < x} -> Int"
x:{x##0 : Int | 0 < x##0} -> Int
λ> doParse' bareTypeP "" "x:Foo x -> Int"
x:(Foo x) -> Int
```

**Comment by ranjitjhala** [comment]
Fixed by #1077 [ specifically, https://github.com/ucsd-progsys/liquidhaskell/pull/1077/commits/9d1403a968ddaf6342d6c6d1c711f57a7a67a23b#diff-0cdb45ae60104f26deeb16471c90f7aaR172]
